src/monaco.contribution.ts

```diff
@@ -138,20 +178,70 @@ enum ModuleResolutionKind {
 }
 //#endregion
 
-const typescriptDefaults = new LanguageServiceDefaultsImpl(
-	{ allowNonTsExtensions: true, target: ScriptTarget.Latest },
-	{ noSemanticValidation: false, noSyntaxValidation: false });
+const languageDefaultOptions = {
+	javascript: {
+		compilerOptions: { allowNonTsExtensions: true, allowJs: true, target: ScriptTarget.Latest },
+		diagnosticsOptions: { noSemanticValidation: true, noSyntaxValidation: false },
+	},
+	typescript: {
+		compilerOptions: { allowNonTsExtensions: true, target: ScriptTarget.Latest },
+		diagnosticsOptions: { noSemanticValidation: false, noSyntaxValidation: false }
+	}
+}
 
-const javascriptDefaults = new LanguageServiceDefaultsImpl(
-	{ allowNonTsExtensions: true, allowJs: true, target: ScriptTarget.Latest },
-	{ noSemanticValidation: true, noSyntaxValidation: false });
+const languageDefaults: { [name: string]: LanguageServiceDefaultsImpl } = {};
+
+function setupLanguageServiceDefaults(languageId, isTypescript) {
+	const languageOptions = languageDefaultOptions[isTypescript ? "typescript" : "javascript"]
+	languageDefaults[languageId] = new LanguageServiceDefaultsImpl(languageId, languageOptions.compilerOptions, languageOptions.diagnosticsOptions);
+}
+
+setupNamedLanguage({
+	id: 'typescript',
+	extensions: ['.ts', '.tsx'],
+	aliases: ['TypeScript', 'ts', 'typescript'],
+	mimetypes: ['text/typescript']
+}, true, true);
+
+setupNamedLanguage({
+	id: 'javascript',
+	extensions: ['.js', '.es6', '.jsx'],
+	firstLine: '^#!.*\\bnode',
+	filenames: ['jakefile'],
+	aliases: ['JavaScript', 'javascript', 'js'],
+	mimetypes: ['text/javascript'],
+}, false, true);
 
 function getTypeScriptWorker(): Promise<any> {
-	return getMode().then(mode => mode.getTypeScriptWorker());
+	return getLanguageWorker("typescript");
 }
 
 function getJavaScriptWorker(): Promise<any> {
-	return getMode().then(mode => mode.getJavaScriptWorker());
+	return getLanguageWorker("javascript");
+}
+
+function getLanguageWorker(languageName: string): Promise<any> {
+	return getMode().then(mode => mode.getNamedLanguageWorker(languageName));
+}
+
+function getLanguageDefaults(languageName: string): LanguageServiceDefaultsImpl {
+	return languageDefaults[languageName];

```

Comment:
Again, this is assuming that properties don't get renamed.

Reply:
Not sure what you mean

Reply:
`languageDefaults` is an object literal defined with an `interface` with non-string keys.  The TypeScript compiler may rename the keys to something else so referencing the properties with string keys isn't safe.

Reply:
`languageDefaults` is just a map we have full control over. Adding items to it is fully handled by the user, so i don't see how the compiler can interfere with that. See https://github.com/Microsoft/monaco-typescript/pull/30/files/292108c5ee680df61aa87a79319caf4713aa6e36#diff-e54df758d643734a71ab22c92c524518R243